vtk_viewer/src/vtk_utils.cpp

```diff
@@ -394,6 +394,67 @@ vtkSmartPointer<vtkPolyData> estimateCurvature(vtkSmartPointer<vtkPolyData> mesh
   return curvature_filter->GetOutput();
 }
 
+void embedRightHandRuleNormals(vtkSmartPointer<vtkPolyData>& data)
+{
+  LOG4CXX_DEBUG(VTK_LOGGER, "Embedding mesh normals from right hand rule");
+  int size = data->GetNumberOfCells();
+  vtkDoubleArray* cell_normals = vtkDoubleArray::New();
+
+  cell_normals->SetNumberOfComponents(3);
+  cell_normals->SetNumberOfTuples(size);
+
+  // loop through all cells and add cell normals
+  for (int i = 0; i < size; ++i)
+  {
+    vtkCell* cell = data->GetCell(i);
+    if (cell)
+    {
+      // Get all point IDs associated with the given cell
+      vtkIdList* pts = cell->GetPointIds();
+      double norm[3] = { 0, 0, 0 };
+
+      // If there are at least 3 points associated with the cell
+      if (pts->GetNumberOfIds() >= 3)
+      {
+        // Define arrays to store the points
+        double p0[3] = { 1, 0, 0 };
+        double p1[3] = { 0, 1, 0 };
+        double p2[3] = { 0, 0, 1 };
+
+        // Extract points from polydata that are associated with this cell
+        data->GetPoint(pts->GetId(0), p0);
+        data->GetPoint(pts->GetId(1), p1);
+        data->GetPoint(pts->GetId(2), p2);
+
+        // Get vectors obeying RHR
+        double v1[3];
+        double v2[3];
+        for (int ind = 0; ind < 3; ind++)
+        {
+          v1[ind] = p1[ind] - p0[ind];
+          v2[ind] = p2[ind] - p1[ind];
+        }
+        // Normal is v1 x v2
+        norm[0] = v1[1] * v2[2] - v2[1] * v1[2];
+        norm[1] = -1 * (v1[0] * v2[2] - v2[0] * v1[2]);
+        norm[2] = v1[0] * v2[1] - v2[0] * v1[1];
+
+        // Normalize the normals
+        vtkMath::Normalize(norm);
+
+        // set the normal for the given cell
+        cell_normals->SetTuple(i, norm);
+      }
+      else
+      {
+        LOG4CXX_ERROR(VTK_LOGGER, "Could not embed normals");
```

Comment:
If this errors out should it not return false? This way that the calling function could then generate the normals as a last resort.

Reply:
That is just an error for a single triangle. With the current method it is not uncommon to get many cells that don't generate normals successfully for whatever reason. I could return false if you want, but it would seem a bit harsh to say the process failed if one triangle was malformed.

Reply:
Could we make it into a warning that prints just once if at least one bad triangle/cell was found?.  You could print more detail on the individual bad cells as a debug message.